lib/replica.rb

```diff
@@ -79,179 +85,181 @@ def self.parse_revision(revision)
+  #   http://tools.wmflabs.org/wikiedudashboard/revisions.php?usernames_User%27&usernames_ids[2]=%27Sage%20(Wiki%20Ed)%27&start=20150105&end=20150108
+  #
+  # Example revisions.php parsed response:
+  # [{"page_id"=>"44962463",
+  #   "page_title"=>"Swarfe/ENGL-122-2014",
+  #   "page_namespace"=>"2",
+  #   "rev_id"=>"641297913",
+  #   "rev_timestamp"=>"20150106205746",
+  #   "rev_user_text"=>"Sage (Wiki Ed)",
+  #   "rev_user"=>"21515199",
+  #   "new_article"=>"false",
+  #   "byte_change"=>"38"},
+  #  {"page_id"=>"44962463",
+  #   "page_title"=>"Swarfe/ENGL-122-2014",
+  #   "page_namespace"=>"2",
+  #   "rev_id"=>"641298113",
+  #   "rev_timestamp"=>"20150106205902",
+  #   "rev_user_text"=>"Sage (Wiki Ed)",
+  #   "rev_user"=>"21515199",
+  #   "new_article"=>"false",
+  #   "byte_change"=>"-50"
+  #  }]
+  def api_get(endpoint, query='')
+    tries ||= 3
+    url = compile_query_url(endpoint, query)
+    response = Net::HTTP::get(URI.parse(url))
+    return unless response.length > 0
+    parsed = JSON.parse response.to_s
+    parsed['data']
+  rescue StandardError => e
+    tries -= 1
+    unless tries.zero?
+      sleep 2
+      retry
     end
+    report_exception e, endpoint, query
+  end
 
-    def compile_query_url(endpoint, query, language=nil)
-      language ||= ENV['wiki_language']
-      base_url = 'http://tools.wmflabs.org/wikiedudashboard/'
-      raw_url = "#{base_url}#{endpoint}?lang=#{language}&#{query}"
-      URI.encode(raw_url)
-    end
+  def compile_query_url(endpoint, query)
+    base_url = 'http://tools.wmflabs.org/wikiedudashboard/'
+    raw_url = "#{base_url}#{endpoint}?lang=#{@wiki.language}&project=#{@wiki.project}&#{query}"
+  end
 
-    # Compile a user list to send to the replica endpoint, which might look
-    # something like this:
-    # "user_ids[0]='Ragesoss'&user_ids[1]='Sage (Wiki Ed)'"
-    def compile_user_ids_string(users)
-      user_list = ''
-      users.each_with_index do |user, i|
-        fail unless user.id
-        user_list += '&' if i > 0
-        user_list += "user_ids[#{i}]='#{user.id}'"
-      end
-      user_list
-    end
+  # Compile a user list to send to the replica endpoint, which might look
+  # something like this:
+  # "usernames[]='Ragesoss'&usernames[]='Sage+%28Wiki+Ed%29'"
+  def compile_usernames_string(users)
+    # FIXME: don't quote
+    { usernames: users.map(&:wiki_id).map {|s| "'#{s}'"} }.to_query
+  end
 
-    def compile_oauth_tags
-      tag_list = ''
-      oauth_ids = ENV['oauth_ids']
-      return '' if oauth_ids.nil?
-      oauth_ids.split(',').each_with_index do |id, i|
-        tag_list += '&' if i > 0
-        tag_list += "oauth_tags[#{i}]='OAuth CID: #{id}'"
-      end
-      tag_list
-    end
+  # Compile a user list to send to the replica endpoint, which might look
+  # something like this:
+  # "user_ids[0]='Ragesoss'&user_ids[1]='Sage (Wiki Ed)'"
+  # FIXME: deprecated
+  def compile_user_ids_string(users)
+    { user_ids: users.map(&:id) }.to_query
```

Comment:
This is awesome. I didn't know about `.to_query`, but it makes this so much less painful!